### PR TITLE
feat(msgpack_rpc): add a new `msgpack-rpc` client type to fix behavior with MessagePack-RPC compliant clients

### PR DESCRIPTION
**Pull request by lambdalisue**
The existing implementation only checks the last frame of the call stack, but the MessagePack-RPC specification does not guarantee the order of messages, so in rare cases message reversals occur and the following error occurs.

    ch 1 returned a response with an unknown request id. Ensure the
    client is properly synchronized

To prevent this, the implementation has been modified to search the entire call stack for the frame corresponding to the message. In order to follow the existing implementation, the call stack is referenced in reverse order.

## Reproduce

It's a bit difficult to reproduce the issue with raw Neovim RPC so we use [denops.vim](https://github.com/vim-denops/denops.vim).

First, clone the following into `/path/to/vim-denops` direcotry

- https://github.com/vim-denops/denops.vim
- https://github.com/vim-denops/denops-issue268

Then write vimrc as

```vim
set rtp+=/path/to/vim-denops/denops.vim
set rtp+=/path/to/vim-denops/denops-issue268
```

Then start Neovim with `nvim --clean -u vimrc` and call `DenopsIssue268` command.

##### Before

https://github.com/neovim/neovim/assets/546312/43c33547-2848-4427-8bed-e73cd82cbf7b

##### After

https://github.com/neovim/neovim/assets/546312/cb499565-5baa-42d0-8a19-62fda2df4b65

## Related

- https://github.com/vim-denops/denops.vim/issues/268

### CODE REVIEW

File: src/nvim/msgpack_rpc/channel.c
Comment (justinmk) [inline src/nvim/msgpack_rpc/channel.c:362]: nit
```suggestion

end:
```

**Review by justinmk — DISMISSED** [review]
> The existing implementation only checks the last frame of the call stack, but the MessagePack-RPC specification does not guarantee the order of messages

`:help msgpack-rpc` documents this Nvim deviation from the msgpack RPC spec:

> Responses must be given in reverse order of requests.

Would be nice to remove that constraint if we can, especially if the fix is as small as this PR suggests. cc @bfredl 

But could this break plugins? Nvim's current constraint presumably makes plugins much simpler because they don't have to deal with out-of-order responses.

## Proposal

- what client is sending the out-of-order responses? can it be fixed?
- alternatively: Nvim could "collect" out-of-order responses and _wait for_ the "missing" response, and only then continue.
    - this will require changes to your PR.

## Related:

- https://github.com/neovim/neovim/issues/20849
- https://github.com/neovim/neovim/issues/19932

**Comment by KillTheMule** [comment]
> Would be nice to remove that constraint if we can,

Does that make sense though? The constraint does not make sense for sequential requests (they must be answered in the same sequence anyways, since neovim blocks on a request, and I don't see this being negotiable), and requests can't "truly" overlap, but only nest. And nested requests need to be answered in reverse order to make any sense. But maybe I'm misunderstanding something :)

**Comment by KillTheMule** [comment]
> Yes. But does my proposal make sense:
> 
> > Nvim could "collect" out-of-order responses and wait for the "missing" response, and only then continue.

A bit, but I have a hard time imagining a properly working plugin that does something like this. I wasn't really able to understand the denops.nvim issue though.

**Comment by lambdalisue** [comment]
> But could this break plugins? Nvim's current constraint presumably makes plugins much simpler because they don't have to deal with out-of-order responses.

I think we will not break it. Because the plugin that changes behavior with this change should now be unusable with the error "ch * returned a response with an unknown requestid".

Also, you say that the current restrictions are simplifying the plugins, but I feel that they are making it even more difficult, since the plugins rather need to return responses to Neovim in order (since normal msgpack RPC would just return them in the order in which they were processed).

> Does that make sense though? The constraint does not make sense for sequential requests (they must be answered in the same sequence anyways, since neovim blocks on a request, and I don't see this being negotiable), and requests can't "truly" overlap (for the same reason), but only nest. And nested requests need to be answered in reverse order to make any sense. But maybe I'm misunderstanding something :)

I think it is true that if a Peer calls multiple Neovim processes, they are returned in order. I think it is a little different because the problem in this case is that when Neovim reads multiple processes of Peer, the responses are not returned in the order in which they were read.

If Neovim is using msgpack RPC, it should be able to handle the responses even if they are not in order.

**Comment by Shougo** [comment]
Current neovim RPC clients(go, python) seems not msgpack RPC incompatible.

Because they must support in order the RPC.  It is hard.

**Comment by peacock0803sz** [comment]
I'll be trying this patch for a while, as it seems to improve the behavior of the plugins I use 👀

**Comment by lambdalisue** [comment]
> Did not see an answer to my suggestion (for this PR), does it make sense or not?:

I'm a little unsure if I'm understanding the suggestions correctly, so could you be more specific?

**Comment by justinmk** [comment]
Currently this PR just looks for a matching frame and processes it. 

My suggestion, instead, is to wait for the "correctly ordered" response. If any out-of-order responses are received, store them for later; don't process them out-of-order. 

However, maybe we should try the "simple" way (this PR) first.

**Comment by lambdalisue** [comment]
Got it thanks :+1: But I don't understand the motivation of the suggesion...

> My suggestion, instead, is to wait for the "correctly ordered" response. If any out-of-order responses are received, store them for later; don't process them out-of-order.

This PR is a change that Neovim supports out-of-order responses, not a change that Neovim returns out-of-order responses. Therefore, I personally don't understand the motivation to reorder out-of-order responses and treat them as in-order responses.

> However, maybe we should try the "simple" way (this PR) first.

Yes, I personally think it would be cost-effective to put it in HEAD once and see if regressions, etc. occur.

File: src/nvim/msgpack_rpc/channel.c
Comment (lambdalisue) [inline src/nvim/msgpack_rpc/channel.c:348]: I think we should just ignore this case while it is likely caused by duplicated messages.

Comment (justinmk) [reply src/nvim/msgpack_rpc/channel.c:348]: Duplicated messages, out-of-order messages... aren't those an indication that something is wrong? If so, silently ignoring them makes things worse. 

Why are messages being duplicated? Should we not at least _log_ such a case?

**Comment by justinmk** [comment]
Looking at https://github.com/vim-denops/denops-issue268 , could a test be fashioned by a single-file deno (without denops) script using Promise.all()? That is much better than nothing.

**Comment by lambdalisue** [comment]
> Looking at https://github.com/vim-denops/denops-issue268 , could a test be fashioned by a single-file deno (without denops) script using Promise.all()? That is much better than nothing. In CI we could install deno.

OK, I'll try but it's a bit difficult so I cannot promise to provide such a sample...

File: src/nvim/msgpack_rpc/channel.c
Comment (lambdalisue) [reply src/nvim/msgpack_rpc/channel.c:348]: > Why are messages being duplicated? Should we not at least log such a case (perhaps at WARN level)? Silently ignoring problems leads to very difficult bugs.

Totally agree. Yes, I said "just ignore" but I didn't mean just ignore, of course 🙇

**Comment by bfredl** [comment]
> The existing implementation only checks the last frame of the call stack, but the MessagePack-RPC specification does not guarantee the order of messages, so in rare cases message reversals occur and the following error occurs.

Note that the root issue remains that the RPC model nvim uses is fundamentally different than the one specified in that spec (stacked layered-cake bi-directional requests with precise stack unwinding vs serial uni-directional requests with out-of-order replies).  This change might make it it easier to get started using some pre-exising msgpack-rpc client stack (although lead to different surprises down the way), but this error message has been quite helpful when you are constructing an RPC client which is fully adapted to nvim's model. 

Regardless, the effect is just that these messages are left buffered until the correct reply has been submitted later, as the top frame of the stack is the only one which nvim is actively waiting on being replied to.

**Comment by justinmk** [comment]
Ok, to unblock this: we need a way to configure this per-channel. That could be either

- a special flag in `nvim_set_client_info()` (I guess a new "attribute", or maybe a special "type"), or
- introduce `nvim_chan_set({chan_id}, {opt})`

In the future we can add other per-channel behaviors, maybe even "fully msgpack-rpc compliant" behavior (per-channel).

**Comment by lambdalisue** [comment]
It would be nice to make it per-channel, but how about introducing `rpcmode({channel}, {mode: 'neovim-rpc' | 'msgpack-rpc'})` or `rpcoptions({channel}, {opt})`? The name `nvim_chan_set` seems too generic for this purpose, since the channel may not be for RPC. Furthermore, we already have `rpcrequest({channel}, ...) ` and `rpcnotify({channel}, ...) `, so `rpcXXXX` would be similar to the existing names and easier to recognize?

**Comment by lambdalisue** [comment]
Note: I'll implement `nvim_set_client_info` with `remote_msgpack`  type on weekend.

**Comment by lambdalisue** [comment]
@justinmk I've added a new type "msgpack-rpc` so please review it

**Comment by lambdalisue** [comment]
Oops. API docs failed. I'll fix thouse and re-request review later.

**Comment by justinmk** [comment]
> The name `nvim_chan_set` seems too generic for this purpose, since the channel may not be for RPC.

The idea is that `nvim_chan_set` would be able to set various options, not only RPC options. 

> we already have `rpcrequest({channel}, ...) ` and `rpcnotify({channel}, ...) `

Those things are just wrappers around the API.

File: src/nvim/msgpack_rpc/channel.c
Comment (justinmk) [inline src/nvim/msgpack_rpc/channel.c:342]: Good instinct! Thank you for enhancing the message.
```suggestion
                 "ch %" PRIu64 " (type=%s) returned a response with an unknown request "
```

File: src/nvim/msgpack_rpc/channel.c
Comment (lambdalisue) [inline src/nvim/msgpack_rpc/channel.c:745]: I'll make it `"remote"` to avoid Segmantation fault.

File: src/nvim/msgpack_rpc/channel.c
Comment (lambdalisue) [inline src/nvim/msgpack_rpc/channel.c:735]: I'll make it `"remote"` to avoid Segmantation fault.

File: src/nvim/msgpack_rpc/channel.c
Comment (justinmk) [inline src/nvim/msgpack_rpc/channel.c:336]: nit: can use strequal
```suggestion
      ChannelCallFrame *frame = strequal(client_type, "msgpack-rpc")
```

Comment (bfredl) [inline src/nvim/msgpack_rpc/channel.c]: seems weird to do string parsing at _this_ point in the codebase. I think we should use an enum, and parse the string right after the client calls `nvim_set_client_info()` .

File: src/nvim/msgpack_rpc/channel.c
Comment (bfredl) [inline src/nvim/msgpack_rpc/channel.c:336]: seems weird to do string parsing at this point in the codebase. I think we should use an enum, and parse the string right after the client calls nvim_set_client_info() .

File: src/nvim/msgpack_rpc/channel.c
Comment (justinmk) [inline src/nvim/msgpack_rpc/channel.c:732]: Almost identical to `rpc_client_name`. Please avoid duplicate code. This can easily be generalized:

```c
const char *get_client_info(Channel *chan, const char *key)
  FUNC_ATTR_NONNULL_ALL
{
  if (!chan->is_rpc) {
    return NULL;
  }
  Dictionary info = chan->rpc.info;
  for (size_t i = 0; i < info.size; i++) {
    if (strequal(key, info.items[i].key.data)
        && info.items[i].value.type == kObjectTypeString) {
      return info.items[i].value.data.string.data;
    }
  }

  return NULL;
}
```

File: src/nvim/msgpack_rpc/channel.c
Comment (lambdalisue) [reply src/nvim/msgpack_rpc/channel.c:336]: It is unlikely, but is it OK if some existing clients use an unknown type, which would break backward compatibility?

File: src/nvim/msgpack_rpc/channel.c
Comment (lambdalisue) [reply src/nvim/msgpack_rpc/channel.c:732]: I think `get_client_info_string_field` would be a better name, since there are non-string fields in `info`. What do you think?

File: src/nvim/msgpack_rpc/channel.c
Comment (justinmk) [reply src/nvim/msgpack_rpc/channel.c]: ? the caller can check for NULL.

File: src/nvim/msgpack_rpc/channel.c
Comment (justinmk) [reply src/nvim/msgpack_rpc/channel.c:336]: the enum can have a `unknown` item

File: src/nvim/msgpack_rpc/channel.c
Comment (justinmk) [reply src/nvim/msgpack_rpc/channel.c:732]: There are no callers using this for anything except string. A fully generalized approach can come later. Most important for now is to avoid duplicating code everywhere, that makes the future harder.

File: src/nvim/msgpack_rpc/channel.c
Comment (lambdalisue) [reply src/nvim/msgpack_rpc/channel.c]: Yes. I simply thought that `type` need not be NULL since the help says that the default value of `type` is `remote`.

File: runtime/doc/api.txt
Comment (justinmk) [inline runtime/doc/api.txt:1344]: This seems like an ok compromise for now. In the future we can revisit `nvim_chan_set` if needed.

File: src/nvim/msgpack_rpc/channel.c
Comment (lambdalisue) [reply src/nvim/msgpack_rpc/channel.c:336]: OK. I'll make it.

**Comment by lambdalisue** [comment]
@justinmk @bfredl I'd like to organize the commits later, but for the time being, could you please review if your request has been met? Or shall I reorganize commits (with force-push) first?

File: runtime/doc/api.txt
Comment (justinmk) [inline runtime/doc/api.txt:1342]: I'm working on updating `:help msgpack-rpc` but let's link to it meanwhile.
```suggestion
                      • "remote" remote client connected "Nvim flavored"
                        MessagePack-RPC (responses must be in reverse order of
                        requests).
```

File: runtime/doc/api.txt
Comment (justinmk) [inline runtime/doc/api.txt:1341]:
```suggestion
                      • "remote" remote client connected "Nvim flavored"
                        MessagePack-RPC (responses must be in reverse order of
                        requests). |msgpack-rpc|
```

**Comment by lambdalisue** [comment]
Using `enum` breaks tests. So should I fix tests or try to make it backward compatible? (convert integer to string on `rpc_client_info`?)

CleanShot 2023-08-07 at 02 26 56

**Comment by justinmk** [comment]
> or try to make it backward compatible? (convert integer to string on rpc_client_info?)

Yes. The enum is only for internal use, to avoid parsing. The parsing should happen _once_, not on every single request.

**Comment by lambdalisue** [comment]
It's kind a off-topic but `cmake --build build --target functionaltest` always freeze after T6425 on my environment (macOS Ventura 13.4.1 on Apple M1 Max). Is there anything I need to install for running tests? I could not find information on Wiki...

CleanShot 2023-08-07 at 02 51 28

**Comment by justinmk** [comment]
> Is there anything I need to install for running tests? I could not find information on Wiki...

no, that's just a bug in the tests.

**Comment by lambdalisue** [comment]
README! I didn't notice that thanks 👍

**Comment by lambdalisue** [comment]
Ah, it seems I misunderstood what was essentially required. The current implementation is inefficient because it scans through the info for each request, so I'll cache this area or something.

**Comment by lambdalisue** [comment]
@justinmk @bfredl I think I finally understood what was essentially required. I've force-pushed it to incorporate all the points you've made so far, so please re-review it :bow:

**Comment by lambdalisue** [comment]
@justinmk @bfredl 
I don't mean to bother you, but I pinged you just in case you missed the notifications (because I often miss notifications).

**Comment by lambdalisue** [comment]
@justinmk @bfredl 

Sorry, but it's been almost two weeks so I mentioned again. Is there anything preventing this PR?

**Comment by bfredl** [comment]
@lambdalisue looks alright. just needs a rebase.

**Comment by lambdalisue** [comment]
@bfredl rebase and force pushed.

**Comment by lambdalisue** [comment]
Thanks 🎉

**Pull request merged**
